### ISSUE TITLE
Troubleshooting: E.164 conversion woes.

### ISSUE THREAD

**Issue by DFurnes** [DoSomething/northstar#637]
It turns out there were some unexpected issues when converting numbers to E.164 format:

### What Happened
1. Capistrano deploys timed out when running the migration on production. This happened when testing the deploy on Thor as well, but it seemed to complete successfully on the following run so I assumed it was a fluke (as we've had other Capistrano deploys randomly time out in the past).
2. This caused the [final step](https://github.com/DoSomething/northstar/blob/e75920c4ca7be9d65c5b69ca178ae0781a7a6cab/database/migrations/2017_09_19_204233_SwapE164ForMobileField.php#L23-L25) of re-adding the unique index on the newly transformed `mobile` field to not occur (although the rename finished as expected, since Mongo continued running the query even after Cap died).
3. Because the index was not re-added, anything that needed to look up a mobile number (like logging in via a mobile, or searching for a user on Rogue) would time out. It seems like these timeouts caused the server to get confused and give `502 Bad Gateway` errors until restarted.
4. Adding a sparse (non-unique) index via the Mongo CLI stopped these issues from occurring.

### Where We Are
It turns out in both Thor & production, we ended up with duplicates after having transformed numbers to E.164 format (since numbers that had previously co-existed as `122233355555` and `22233355555` both got transformed into the same E.164 format). (#410, #415)

I didn't catch this since the `e164` temporary field we were using didn't get a unique index, and I only checked that the total counts were consistent between the two (and a few data spot-checks here and there).

### What's Next
Everything seems to have stabilized after adding that sparse index, but we'll now have weird behavior when trying to look up any of those "dupe" users. I'm going to make a quick script to fix those records by discarding the records with the leading `1##########` on their `_old_mobile` & we can then re-add the unique index on the `mobile` field.

**Comment by DFurnes** [comment]
Merging #638 and going to test it out on QA/Thor. If all seems good, will push to prod as well.

**Comment by DFurnes** [comment]
I tried running the de-dupe script on Thor, and am getting a MongoDB fatal error:

```sh
$ php artisan northstar:e164-dupes --pretend


  [MongoDB\Driver\Exception\RuntimeException]
  BSONObj size: 63913327 (0x3CF3D6F) is invalid. Size must be between 0 and 16793600(16MB)
```

It seems like this error occurs when indexes get corrupted (which could totally make sense since the database migration got cut off partway through). The posts I'm seeing online suggest running [`db.repairDatabase()`](https://docs.mongodb.com/manual/reference/command/repairDatabase/#dbcmd.repairDatabase) from the REPL... although that's apparently a bad idea™ if using a replica set and we should [re-sync the data from an intact member instead](https://docs.mongodb.com/manual/tutorial/resync-replica-set-member/). 🤔

This is getting out of my depth, so pinging @blisteringherb @sheyd as the Mongo experts.

**Comment by DFurnes** [comment]
From what I can tell by running the [`rs.status()` command](https://docs.mongodb.com/manual/reference/command/replSetGetStatus/#rs-status-output) on the Thor cluster, each of the nodes seem to be healthy so not 100% sure that's it. Definitely open to any other ideas for what could be wrong here.

**Comment by DFurnes** [comment]
Oy, it wasn't that at all! It seems that the query was trying to bundle up all the "duplicates" of records that didn't have a `mobile` set and since there's a ton like that, it was running out of memory. The fix in #639 should get us going again.

**Comment by DFurnes** [comment]
After merging #639, I was able to run this script on both QA (where it removed 1 dupe) & Thor (where it removed 1,228). Both have log output stored at `~/e164-dupes.log` for future reference.

Re-creating the unique index on the `mobile` field now works as expected. 🥞

```js
db.users.dropIndex('mobile_1');
{ "nIndexesWas" : 9, "ok" : 1 }

db.users.createIndex({mobile: 1}, {unique: true, sparse: true});
{
        "createdCollectionAutomatically" : false,
        "numIndexesBefore" : 8,
        "numIndexesAfter" : 9,
        "ok" : 1
}
```

Will run on production in the morning.

**Comment by DFurnes** [comment]
And this is done! Removed the 1,228 duplicates on production, and recreated the unique index. ✅ 

Going to audit whether any other indexes were improperly created due to 2e565d1.

**Comment by DFurnes** [comment]
And the other indexes all look great. 👌

A few things we can probably learn from this:

 - Let's consider increasing the Capistrano timeout, so we avoid more incomplete deploys when a migration takes longer than expected (maybe 10 minutes?).
 - Let's try to remember to put sites into maintenance mode with `php artisan down` whenever we're deploying a heftier migration, so that users receive a "Be back soon." & downstream services receive a 503 instead of potentially incorrect data.
 - Other services didn't seem to handle Northstar's 502 errors all that gracefully, so let's consider running a fire drill in Thor sometime to suss out where the pain points are and try to add more graceful error handling.
 - Let's more carefully audit the commits included in each release of [laravel-mongodb](https://github.com/jenssegers/laravel-mongodb) so we don't get any more surprise changes that aren't in the release notes.

**Comment by DFurnes** [comment]
And one last issue: @jamjensen reported that search results stopped working after this deploy, and it turned out this was caused by normalizing invalid mobiles to `null` instead of an empty string. I added a test case & fix in #640.
